### PR TITLE
Update Pow and Slice OP.

### PR DESCRIPTION
**Pull request by ebarsoum**
The exponent of pow OP need to be an input and I updated slice slightly.

### CODE REVIEW

File: onnx/defs/tensor/defs.cc [hunk 1]
```diff
@@ -74,7 +74,7 @@ to equal sized parts.
 )DOC");
 
 OPERATOR_SCHEMA(Slice)
-    .NumInputs(1, 3)
+    .NumInputs(1)
```
Comment (dzhulgakov) [inline onnx/defs/tensor/defs.cc]: actually C2 implementation allows to specify starts and ends as arguments or inputs. Thus support for (3) here. @ezyang - do we rely on it anywhere?

File: onnx/defs/math/defs.cc [hunk 2]
```diff
@@ -259,19 +259,17 @@ and output blobs.
...
 
 OPERATOR_SCHEMA(Pow)
-    .NumInputs(1)
+    .NumInputs(2)
     .NumOutputs(1)
-    .Attr("exponent",
-          "The exponent of the power function.",
-          AttrType::FLOAT)
-    .AllowConsumed({{0, 0}})
+    .AllowConsumed({{0, 0}, {1, 0}})
```
Comment (dzhulgakov) [inline onnx/defs/math/defs.cc]: probably we don't want to allow for the second argument to be consumed

**Comment by dzhulgakov** [comment]
What's the general procedure for updating the op specs? It means that we'd need to change backend implementations (like in Pow case)

File: onnx/defs/math/defs.cc [hunk 3]
```diff
@@ -259,19 +259,17 @@ and output blobs.
                "computed element-wise");
 
 OPERATOR_SCHEMA(Pow)
-    .NumInputs(1)
+    .NumInputs(2)
     .NumOutputs(1)
-    .Attr("exponent",
```
Comment (bddppq) [inline onnx/defs/math/defs.cc]: I would prefer keeping this attribute. Frontends could already know the exponent and it will be inconvenient (also inefficient) to always create an extra input tensor).

File: onnx/defs/tensor/defs.cc [hunk 4]
```diff
@@ -89,6 +89,9 @@ the end of that dimension.
 reverse order.
 )DOC")
     .Input(0, "data", "Tensor of data to extract slices from.")
+    .Attr("axes",
```
Comment (bddppq) [inline onnx/defs/tensor/defs.cc]: Just to make sure I understand this, so `axes=[1, 0]` means `data[starts[1]:ends[1], starts[0]:ends[0]]`? If so, since these are attributes, so I assume they are already known at the time frontends export onnx, why couldn't frontends reshuffle starts and ends?

File: onnx/defs/tensor/defs.cc (hunk 1 above)
Comment (ebarsoum) [reply onnx/defs/tensor/defs.cc]: Why do we need both: attribute and input?

File: onnx/defs/math/defs.cc (hunk 2 above)
Comment (ebarsoum) [reply onnx/defs/math/defs.cc]: Sure

File: onnx/defs/math/defs.cc (hunk 3 above)
Comment (ebarsoum) [reply onnx/defs/math/defs.cc]: The problem is we have case in which the exponent can change, specially for custom learner. We added exponent as input in order to support Keras backend and other scenario. Atteribute will be limiting in this case.

File: onnx/defs/tensor/defs.cc (hunk 4 above)
Comment (ebarsoum) [reply onnx/defs/tensor/defs.cc]: This isn't for shuffle, actually the above won't work. Let's say you have 5D tensor as input, and you want to Slice the 2 in the middle, you can say: axes = (2,3), start=(1,1), end=(4,4). So axes, specify the list of axes in which you apply the start and end indices, len(axes) = len(starts) = len(ends). This is better, then specifying all the axes in starts and ends, and most of the starts and ends are the entire dimension.

File: onnx/defs/math/defs.cc (hunk 3 above)
Comment (bddppq) [reply onnx/defs/math/defs.cc]: I'm not against adding exponent as Input for dynamic cases, I'm saying let's keep the attribute for static cases.

File: onnx/defs/tensor/defs.cc (hunk 4 above)
Comment (bddppq) [reply onnx/defs/tensor/defs.cc]: I see that makes senses

File: onnx/defs/math/defs.cc (hunk 3 above)
Comment (ebarsoum) [reply onnx/defs/math/defs.cc]: @bddppq  what the advantage? static case could use the argument also? What if the user specify both?

Comment (bddppq) [reply onnx/defs/math/defs.cc]: Copy from my first comment:

> Frontends could already know the exponent and it will be inconvenient (also inefficient) to always create an extra input tensor).

In caffe2 we (will) disallow user specifying both.

File: onnx/defs/tensor/defs.cc (hunk 1 above)
Comment (bddppq) [reply onnx/defs/tensor/defs.cc]: The reason is exactly like adding `exponent` as input in Pow op: for dynamic use cases :-)

File: onnx/defs/tensor/defs.cc (hunk 4 above)
Comment (dzhulgakov) [reply onnx/defs/tensor/defs.cc]: Another way is to allow -1 to denote "the rest" like in Python and always specify the sizes. TF does the same: https://www.tensorflow.org/api_docs/python/tf/slice

Same with C2: https://github.com/caffe2/caffe2/blob/01827c153db96349745a544148c1ff0386c5ef9e/caffe2/operators/slice_op.h#L59

I feel it's more natural. What do you think?

File: onnx/defs/tensor/defs.cc (hunk 1 above)
Comment (ebarsoum) [reply onnx/defs/tensor/defs.cc]: It look weird to me to have one function support both instead of split them, and not all OPs require dynamic input. If that the case, we can promote all our attributes to input.s

File: onnx/defs/tensor/defs.cc (hunk 4 above)
Comment (ebarsoum) [reply onnx/defs/tensor/defs.cc]: -1 mean read from the end.

File: onnx/defs/tensor/defs.cc (hunk 1 above)
Comment (ezyang) [reply onnx/defs/tensor/defs.cc]: PyTorch's export which uses slice always uses slice's inputs as constants, and we currently fake up a tensor in order to support it. There are other operators that PyTorch supports which would make use of the dynamic behavior.

Comment (fmassa) [reply onnx/defs/tensor/defs.cc]: Note that TF uses inputs for both `Transpose` and `Reshape` instead of attributes for the permutation / new shape.

Comment (bddppq) [reply onnx/defs/tensor/defs.cc]: @ebarsoum Since there will be usecases that needs `starts` and `ends` to be dynamic, we need to add them back to Input (and remove them from attributes). Since "axes" is tied to "starts" and "ends', I believe it should be converted to input as well.

**Comment by ebarsoum** [comment]
I updated this PR, based on all feed backs. Can I get a sign off?

**Comment by bddppq** [comment]
How about the "axes", "starts" and "ends" attribute in the Slice op?? Shouldn't they be deleted since now they are inputs?

**Comment by jerryzh168** [comment]
@bddppq I think it's better to remove inputs and keep attributes, what's the argument here to keep/remove either one? I just did a translation of Crop in C1 to Slice, and used attributes because inputs requires the type matching and somehow I didn't manage to get it right.(it's complaining that it was given a long long type but I have explicitly cast it to int), So I think it will be less pain to move these things to attribute.

File: onnx/defs/tensor/defs.cc (hunk 4 above)
Comment (jerryzh168) [reply onnx/defs/tensor/defs.cc]: As an fyi, in c2, slice don't have "axis" and user needs to specify the starts and ends of each dimension, and currently c2 only supports slicing one of the dimensions. We might need to change c2 implementation when this is merged.

**Comment by bddppq** [comment]
@jerryzh168 I actually like to keep both too, it's much more practical for frontends (no need to create extra constant tensor to host values that are already known). But there was an offline discussion that come to the conclusion we should keep only one of input & attribute.
@ezyang @ebarsoum @yuanbyu

File: onnx/defs/tensor/defs.cc (hunk 4 above)
Comment (dzhulgakov) [reply onnx/defs/tensor/defs.cc]: Yeah, I mean that instead of `axis` you can fill other dimensions with [0, -1]

In other words for 5D tensor and axes = (2,3), start=(1,1), end=(4,4) it'd become
`start=(0,0,1,1,0), end=(-1,-1,4,4,-1)` and it's more explicit.

File: onnx/defs/tensor/defs.cc [hunk 5]
```diff
@@ -74,21 +74,24 @@ to equal sized parts.
...
     .NumOutputs(1)
     .SetDoc(R"DOC(
 Produces a slice of the input tensor along multiple axes. Similar to numpy:
 https://docs.scipy.org/doc/numpy/reference/arrays.indexing.html 
 
-Slices are passed as two keyword argument lists with starting and end indices 
-for each dimension of the input `data` tensor. If a negative value is passed 
-for any of the start or end indices, it represent number of elements before 
-the end of that dimension.
+Slices uses `axes`, `starts` and `ends` list to specify the start and end dimension
```
Comment (dzhulgakov) [inline onnx/defs/tensor/defs.cc]: clarify whether the end is inclusive or exclusive?

**Comment by ezyang** [comment]
I summarized the offline discussion, and future things we have to do, here; https://github.com/onnx/onnx/issues/76

File: onnx/defs/tensor/defs.cc (hunk 4 above)
Comment (jerryzh168) [reply onnx/defs/tensor/defs.cc]: Also in numpy/matlab we need to specify all the dimensions in array slicing. I think the definition of c2 slice op is probably borrowed from numpy.
